cdc/kv/regionlock/region_range_lock.go

```diff
@@ -31,84 +31,194 @@ import (
 	ts       uint64
+
+	// rangeTsEntry is used in rangeTsMap. rangeTsMap.set will associate a timestamp to a given key range,
+	// and rangeTsMap.unset will revoke the relationship. `isUnset` indicates whether a key range is
+	// generated from a set or unset operation.
+	isUnset bool
 }
 
-func rangeTsEntryWithKey(key []byte) *rangeTsEntry {
-	return &rangeTsEntry{
-		startKey: key,
-	}
+func rangeTsEntryWithKey(key []byte) rangeTsEntry {
+	return rangeTsEntry{startKey: key}
 }
 
-func rangeTsEntryLess(a, b *rangeTsEntry) bool {
+func rangeTsEntryLess(a, b rangeTsEntry) bool {
 	return bytes.Compare(a.startKey, b.startKey) < 0
 }
 
 // rangeTsMap represents a map from key range to a timestamp. It supports
 // range set and calculating min value among a specified range.
 type rangeTsMap struct {
-	m *btree.BTreeG[*rangeTsEntry]
+	m     *btree.BTreeG[rangeTsEntry]
+	start []byte
+	end   []byte
 }
 
 // newRangeTsMap creates a RangeTsMap.
 func newRangeTsMap(startKey, endKey []byte, startTs uint64) *rangeTsMap {
 	m := &rangeTsMap{
-		m: btree.NewG(16, rangeTsEntryLess),
+		m:     btree.NewG(16, rangeTsEntryLess),
+		start: startKey,
+		end:   endKey,
 	}
-	m.Set(startKey, endKey, startTs)
+	m.set(startKey, endKey, startTs)
 	return m
 }
 
-// Set sets the corresponding ts of the given range to the specified value.
-func (m *rangeTsMap) Set(startKey, endKey []byte, ts uint64) {
-	if _, ok := m.m.Get(rangeTsEntryWithKey(endKey)); !ok {
-		// To calculate the minimal ts, the default value is math.MaxUint64
-		tailTs := uint64(math.MaxUint64)
-		m.m.DescendLessOrEqual(rangeTsEntryWithKey(endKey), func(i *rangeTsEntry) bool {
-			tailTs = i.ts
+func (m *rangeTsMap) clone() (res *rangeTsMap) {
+	res = &rangeTsMap{
+		m:     btree.NewG(16, rangeTsEntryLess),
+		start: m.start,
+		end:   m.end,
+	}
+	m.m.Ascend(func(i rangeTsEntry) bool {
+		res.m.ReplaceOrInsert(i)
+		return true
+	})
+	return
+}
+
+func (m *rangeTsMap) set(startKey, endKey []byte, ts uint64) {
+	if !bytes.Equal(m.end, endKey) && !m.m.Has(rangeTsEntryWithKey(endKey)) {
+		var found bool
+		var endKeyOverlapped rangeTsEntry
+		m.m.DescendLessOrEqual(rangeTsEntryWithKey(endKey), func(i rangeTsEntry) bool {
+			found = true
+			endKeyOverlapped.ts = i.ts
+			endKeyOverlapped.isUnset = i.isUnset
 			return false
 		})
-		m.m.ReplaceOrInsert(&rangeTsEntry{
-			startKey: endKey,
-			ts:       tailTs,
-		})
+		if found {
+			if !endKeyOverlapped.isUnset {
+				log.Panic("rangeTsMap double set")
+			}
+			endKeyOverlapped.startKey = endKey
+			m.m.ReplaceOrInsert(endKeyOverlapped)
+		}
 	}
 
-	entriesToDelete := make([]*rangeTsEntry, 0)
-	m.m.AscendRange(rangeTsEntryWithKey(startKey), rangeTsEntryWithKey(endKey),
-		func(i *rangeTsEntry) bool {
-			entriesToDelete = append(entriesToDelete, i)
-			return true
+	if !m.m.Has(rangeTsEntryWithKey(startKey)) {
+		var found bool
+		var startKeyOverlapped rangeTsEntry
+		m.m.DescendLessOrEqual(rangeTsEntryWithKey(startKey), func(i rangeTsEntry) bool {
```

Comment:
Is the core idea here to find the node before the target node?

Reply:
There must be a node overlapped with the target. We need to find and update it.